### PR TITLE
add go extension trailer del api

### PR DESCRIPTION
**Pull request by StarryVae**
Commit Message: add go extension trailer del api
Additional Description: N/A
Risk Level: low
Testing: interation test
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
@doujiang24 PTAL, thanks.

**Review by doujiang24 — COMMENTED** [review]
otherwise, LGTM.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 1]
```diff
@@ -855,6 +855,25 @@ CAPIStatus Filter::setTrailer(absl::string_view key, absl::string_view value, he
...
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+  if (trailers_ == nullptr) {
+    ENVOY_LOG(debug, "invoking cgo api at invalid phase: {}", __func__);
+    return CAPIStatus::CAPIInvalidPhase;
+  }
+  trailers_->remove(Http::LowerCaseString(key));
```
[the remark this answers is not in the document]
Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: done, PTAL again, thanks.

**Review by doujiang24 — APPROVED** [review]

**Comment by StarryVae** [comment]
cc @wbpcode @mattklein123

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by Axkea** [comment]
I think I ran into some trouble when I called log function and it ran straight out of memory

I trace that it called C.GoStringN() and then it reported an error

Here is the stack information for the call
fatal error: runtime: out of memory

runtime stack:
runtime.throw({0x7fcec6a86fb1?, 0x7fcec728ed50?})
	/usr/local/go_sdk/go/src/runtime/panic.go:1047 +0x5f fp=0x7fce9af27bc0 sp=0x7fce9af27b90 pc=0x7fcec65b223f
runtime.sysMapOS(0x7fce98000000, 0xffffffffff800000?)
	/usr/local/go_sdk/go/src/runtime/mem_linux.go:187 +0x11b fp=0x7fce9af27c08 sp=0x7fce9af27bc0 pc=0x7fcec6594c1b
runtime.sysMap(0x7fcec728ed50?, 0x7fcec5ba8000?, 0x1c00000?)
	/usr/local/go_sdk/go/src/runtime/mem.go:142 +0x35 fp=0x7fce9af27c38 sp=0x7fce9af27c08 pc=0x7fcec65945f5
runtime.(*mheap).grow(0x7fcec728ed40, 0x7fffffffffade?)
	/usr/local/go_sdk/go/src/runtime/mheap.go:1468 +0x23d fp=0x7fce9af27ca8 sp=0x7fce9af27c38 pc=0x7fcec65a51dd
runtime.(*mheap).allocSpan(0x7fcec728ed40, 0x7fffffffffade, 0x0, 0x1)
	/usr/local/go_sdk/go/src/runtime/mheap.go:1199 +0x1be fp=0x7fce9af27d40 sp=0x7fce9af27ca8 pc=0x7fcec65a491e
runtime.(*mheap).alloc.func1()
	/usr/local/go_sdk/go/src/runtime/mheap.go:918 +0x65 fp=0x7fce9af27d88 sp=0x7fce9af27d40 pc=0x7fcec65a43a5
runtime.systemstack()
	/usr/local/go_sdk/go/src/runtime/asm_amd64.s:492 +0x46 fp=0x7fce9af27d90 sp=0x7fce9af27d88 pc=0x7fcec65e0286

goroutine 17 [running, locked to thread]:
runtime.systemstack_switch()
	/usr/local/go_sdk/go/src/runtime/asm_amd64.s:459 fp=0xc00108f910 sp=0xc00108f908 pc=0x7fcec65e0220
runtime.(*mheap).alloc(0xffffffffff5bc000?, 0x7fffffffffade?, 0xe0?)
	/usr/local/go_sdk/go/src/runtime/mheap.go:912 +0x65 fp=0xc00108f958 sp=0xc00108f910 pc=0x7fcec65a42e5
runtime.(*mcache).allocLarge(0x7fcec6da1e60?, 0xffffffffff5bbd00, 0x1)
	/usr/local/go_sdk/go/src/runtime/mcache.go:233 +0x85 fp=0xc00108f9a8 sp=0xc00108f958 pc=0x7fcec6593585
runtime.mallocgc(0xffffffffff5bbd00, 0x0, 0x0)
	/usr/local/go_sdk/go/src/runtime/malloc.go:1029 +0x57e fp=0xc00108fa20 sp=0xc00108f9a8 pc=0x7fcec65897de
runtime.rawstring(0xffffffffff5bbd00)
	/usr/local/go_sdk/go/src/runtime/string.go:273 +0x37 fp=0xc00108fa70 sp=0xc00108fa20 pc=0x7fcec65cd437
runtime.gostringn(0x7fcec6e80d80?, 0xffffffffff5bbd00)
	/usr/local/go_sdk/go/src/runtime/string.go:344 +0x2b fp=0xc00108fab0 sp=0xc00108fa70 pc=0x7fcec65cd84b
github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http._Cfunc_GoStringN(...)
	_cgo_gotypes.go:109
github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http.(*httpRequest).pluginName(0xc000385aa0)
	/root/go/pkg/mod/github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http/filter.go:67 +0x6b fp=0xc00108faf8 sp=0xc00108fab0 pc=0x7fcec6a04fcb
github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http.(*httpRequest).Log(0x7fcec6a6b458?, 0x5?, {0xc00177f3b0, 0x2f})
	/root/go/pkg/mod/github.com/envoyproxy/envoy/contrib/golang/filters/http/source/go/pkg/http/filter.go:116 +0x2c fp=0xc00108fb58 sp=0xc00108faf8 pc=0x7fcec6a0548c

**Comment by doujiang24** [comment]
@Axkea Thanks for your report, you'd better report in a new issue, with reproducable minimal example code and steps.

**Pull request merged**